### PR TITLE
Unify Engine API failure responses

### PR DESCRIPTION
**Pull request by rubo**
## Changes

Implemented https://github.com/ethereum/execution-apis/pull/337 that unifies Engine API failure responses.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a fix or a feature that would cause existing functionality not to work as expected)
- [ ] Documentation update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional or API changes)
- [ ] Build-related changes
- [ ] Other: _description_ 

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

Withdrawals Hive tests passing.

### CODE REVIEW

**Review by LukaszRozmej — COMMENTED** [review]
Maybe we should do class hierarchy for `ExecutionPayload`?

File: src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs [hunk 1]
```diff
@@ -72,14 +76,31 @@ public async Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV1(ExecutionP
...
-            NewPayload(executionPayload, nameof(engine_newPayloadV2));
+        public Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV2(ExecutionPayload executionPayload)
+        {
+            var spec = _specProvider.GetSpec((executionPayload.BlockNumber, executionPayload.Timestamp));
+            string? error = null;
+
+            if (spec.WithdrawalsEnabled && executionPayload.Withdrawals is null)
+                error = "ExecutionPayloadV2 expected";
+            else if (!spec.WithdrawalsEnabled && executionPayload.Withdrawals is not null)
+                error = "ExecutionPayloadV1 expected";
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs:87]: Would like to see versioning and validation abstracted at least to the extension methods, to be clearer.

**Comment by rubo** [comment]
> Maybe we should do class hierarchy for ExecutionPayload?
Or at least a clear way of validating them, for example, extension methods like: IsValidV1, IsValidV2 or GetVersion.

That was _exactly_ the way I went initially but it turned out to be a longer and more complicated approach. Class hierarchy won't work in this case either. The only way to distinguish versions, for now, is to check whether the `Withdrawals` property is `null` based on the timestamp and block number combination.
Given this check is _only_ needed at the very beginning of _this particular method_, moving it to an extension method doesn't bring any value here. If we're going to reuse it somewhere else as well, then it would make sense, indeed.

Also, because this check is an input parameter validation, I'd like to have it before any other checks for FCU as well. However, the timestamp and block number logic are a bit complicated there, so I left it there.

Things may change with future version checks though.

**Comment by LukaszRozmej** [comment]
> > Maybe we should do class hierarchy for ExecutionPayload?
> > Or at least a clear way of validating them, for example, extension methods like: IsValidV1, IsValidV2 or GetVersion.
> 
> That was _exactly_ the way I went initially but it turned out to be a longer and more complicated approach. Class hierarchy won't work in this case either. The only way to distinguish versions, for now, is to check whether the `Withdrawals` property is `null` based on the timestamp and block number combination. Given this check is _only_ needed at the very beginning of _this particular method_, moving it to an extension method doesn't bring any value here. If we're going to reuse it somewhere else as well, then it would make sense, indeed.
> 
> Also, because this check is an input parameter validation, I'd like to have it before any other checks for FCU as well. However, the timestamp and block number logic are a bit complicated there, so I left it there.
> 
> Things may change with future version checks though.

My whole point is on semantics and readability lets name it V1/V2 and make validation based on that. Will be easier to read and translate from specs, let me try to provide example

**Comment by LukaszRozmej** [comment]
created versioning PR: https://github.com/NethermindEth/nethermind/pull/5157

**Review by MarekM25 — APPROVED** [review]

**Pull request merged**